module/VuFind/src/VuFind/View/Helper/Root/Record.php

```diff
@@ -707,4 +706,14 @@ protected function deduplicateLinks($links)
     {
         return array_values(array_unique($links, SORT_REGULAR));
     }
+
+    /**
+     * Get the source identifier + unique id of the record
+     *
+     * @return string
+     */
+    public function getUniqueIdWithSource()
```

Comment:
Some further suggested revisions here. Maybe we should rename this to `getUniqueIdWithSourcePrefix` to be very specific about what is happening, and then create a new method below for use in the templates:

```
public function getUniqueHtmlElementId()
{
    return preg_replace("/\s+/", "_", $this->getUniqueIdWithSource());
}
```

After doing a bit of research, I see that HTML5 element IDs support nearly all characters, but not whitespace -- so to be on the safe side, I think when we use a record ID as an element ID, we should ensure that whitespace is eliminated. (Hopefully the odds of an ID collision caused by replacing whitespace with underscores are small enough that we don't need to worry... but if we ARE worried, we could also do something different like `'record_' . md5($this->getUniqueIdWithSourcePrefix())` to more reliably guarantee uniqueness. This would also offer a nice hook if people wanted to implement different patterns for element ID generation in their own code.

Reply:
Changed that. Should we use the escapeHtmlAttr helper directly within the record view helper and remove escaping of in templates?

Reply:
Thanks @RLangeUni! I don't think auto-escaping the ID will work, because in some cases we're passing the ID as an array value to the URL helper, which will also automatically escape it -- and we don't want to end up with double-escaped values. We could add a flag parameter for escaping, or have two different method names for escaped vs. unescaped versions, but that seems potentially overcomplicated. It might be tidier to refactor more of the templates to use the `htmlAttributes()` or `makeTag()` view helpers for constructing tags with lots of attributes from arrays, and then all the escaping is more cleanly built-in and there's less potential confusion about what is or is not escaped. (But that's a project for another day, I think :-) ).